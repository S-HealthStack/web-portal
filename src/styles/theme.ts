import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
