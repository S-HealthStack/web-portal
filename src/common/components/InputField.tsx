import React, { FC, ForwardedRef, forwardRef } from 'react';

import styled, { css } from 'styled-components';

import browser from 'src/common/utils/browser';
import { animation, colors, px, typography } from 'src/styles';

export const RIGHT_PADDING = 8;

export interface InputFieldBaseProps extends React.InputHTMLAttributes<HTMLInputElement> {
  label?: string | JSX.Element;
  helperText?: string;
  error?: boolean | string; // TODO: remove `boolean` when migration to v0.9 is complete
  withoutErrorText?: boolean;
  disabled?: boolean;
}
type InputFieldShellProps = React.PropsWithChildren<InputFieldBaseProps>;

export interface InputFieldProps
  extends InputFieldBaseProps,
    React.InputHTMLAttributes<HTMLInputElement> {
  type?: 'email' | 'password' | 'text' | 'date' | 'number';
  endExtra?: { component: JSX.Element; extraWidth: number };
  lighten?: boolean;
}

export const InputContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: ${px(8)};
`;

export const ExtraWrapper = styled.div<Pick<InputFieldProps, 'endExtra'>>`
  > svg {
    position: relative;
    bottom: ${browser.isSafari ? px(43) : px(41)};
    left: ${({ endExtra }) => `calc(100% - ${px((endExtra?.extraWidth || 0) + RIGHT_PADDING)})`};
  }
`;

const getDefaultBackgroundColor = ({ error, lighten }: InputFieldProps) => {
  if (error) {
    return colors.statusError10;
  }

  return lighten ? colors.surface : colors.background;
};

export const StyledTextField = styled.input<InputFieldProps>`
  ${typography.bodyMediumRegular};
  color: ${({ error }) => (error ? colors.statusErrorText : colors.textPrimary)};
  box-sizing: border-box;
  height: ${px(56)};
  width: 100%;
  margin: 0;
  padding: ${px(16)};
  background-color: ${getDefaultBackgroundColor};
  border: ${px(1)} solid ${getDefaultBackgroundColor};
  border-radius: ${px(4)};
  transition: border 300ms ${animation.defaultTiming};
  caret-color: ${({ error }) => (error ? colors.statusErrorText : colors.textPrimaryBlue)};

  &:hover:enabled {
    border-color: ${({ error }) => (error ? 'transparent' : colors.primaryHovered)};
  }

  &:active:enabled,
  &:focus-visible {
    outline: none;
    border-color: ${({ error }) => (error ? 'transparent' : colors.primary)};
  }

  &:disabled {
    color: ${colors.disabled};
    border-color: ${colors.disabled20};
    background-color: ${colors.disabled20};
  }

  &::placeholder {
    color: ${colors.textSecondaryGray};
  }

  &:disabled::placeholder {
    color: ${colors.disabled};
  }

  &:-webkit-autofill,
  &:-webkit-autofill:hover,
  &:-webkit-autofill:focus {
    ${({ error, disabled, theme }) => css`
      border: ${px(1)} solid
        ${(disabled && theme.colors.disabled20) ||
        (error ? theme.colors.statusError10 : theme.colors.primary)};
      transition: background-color 5000s ease-in-out 0s;
      background-color: ${getDefaultBackgroundColor} !important;
      background-image: unset !important;
      -webkit-text-fill-color: ${colors.textDisabled};
    `};
  }

  &:-webkit-autofill:focus {
    border-color: ${({ error }) => (error ? 'transparent' : colors.primary)};
  }
`;

export const InputWrapper = styled.div<Pick<InputFieldProps, 'endExtra' | 'error'>>`
  max-height: ${px(56)};
  &:hover {
    ${StyledTextField} {
      :enabled {
        border-color: ${({ error }) => (error ? 'transparent' : colors.primaryHovered)};
      }
    }
  }
`;

interface BlockStatus {
  disabled?: boolean;
  error?: boolean;
}

export const Label = styled.div<BlockStatus>`
  ${typography.bodyMediumSemibold};
  color: ${({ error }) => (error ? colors.statusErrorText : colors.textPrimary)};
  height: ${px(18)};
`;

export const InputDescription = styled.div<BlockStatus>`
  ${typography.bodySmallRegular};
  color: ${({ error }) => (error ? colors.statusErrorText : colors.textPrimary)};
  gap: ${px(8)};
  height: ${px(18)};
`;

export const InputErrorText = styled.div<{ withOffset?: boolean }>`
  ${typography.bodySmallRegular};
  color: ${colors.statusErrorText};
  padding-left: ${({ withOffset }) => withOffset && px(16)};
  height: ${px(17)};
`;

export const InputFieldShell: FC<InputFieldShellProps> = ({
  label,
  helperText,
  error,
  className,
  disabled,
  children,
  withoutErrorText,
}) => (
  <InputContainer className={className}>
    {helperText && (
      <Label data-testid="input-label" error={!!error} disabled={disabled}>
        {label}
      </Label>
    )}
    <InputDescription data-testid="input-description" error={!!error} disabled={disabled}>
      {helperText || label || <>&nbsp;</>}
    </InputDescription>
    <InputWrapper error={error}>{children}</InputWrapper>
    {!withoutErrorText && (
      <InputErrorText data-testid="input-error" withOffset={!helperText}>
        {typeof error === 'string' ? error : <>&nbsp;</>}
      </InputErrorText>
    )}
  </InputContainer>
);

const InputField = forwardRef(
  (
    {
      type,
      helperText,
      error,
      label,
      endExtra,
      disabled,
      className,
      withoutErrorText,
      ...restProps
    }: InputFieldProps,
    ref: ForwardedRef<HTMLInputElement>
  ): JSX.Element => (
    <InputFieldShell
      label={label}
      helperText={helperText}
      error={error}
      className={className}
      disabled={disabled}
      withoutErrorText={withoutErrorText}
    >
      <StyledTextField
        data-testid="input"
        ref={ref}
        error={error}
        type={type}
        disabled={disabled}
        endExtra={endExtra}
        {...restProps}
      />
      <ExtraWrapper endExtra={endExtra}>{endExtra?.component}</ExtraWrapper>
    </InputFieldShell>
  )
);

export default InputField;
