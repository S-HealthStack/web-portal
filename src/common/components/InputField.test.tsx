import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import InputField from 'src/common/components/InputField';
import theme from 'src/styles/theme';
import { ThemeProvider } from 'styled-components/';

describe('InputField', () => {
  it('type and clear InputField test', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <InputField />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const input = screen.getByTestId('input');

    await userEvent.type(input, 'Samsung');
    expect(input).toHaveValue('Samsung');

    await userEvent.clear(input);
    expect(input).toHaveValue('');
  });

  it('test InputField error state', () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <InputField error />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const input = screen.getByTestId('input');

    expect(input).toHaveStyle(`color: ${theme.colors.statusErrorText}`);
    expect(input).toHaveStyle(`caret-color: ${theme.colors.statusErrorText}`);
  });

  it('test InputField disabled state', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <InputField disabled value="disabled" />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const input = screen.getByTestId('input');

    await userEvent.type(input, 'Samsung');
    expect(input).toHaveValue('disabled');
    expect(input).toHaveStyle(`color: ${theme.colors.disabled}`);
    expect(input).toHaveStyle(`background-color: ${theme.colors.disabled20}`);
  });
});
