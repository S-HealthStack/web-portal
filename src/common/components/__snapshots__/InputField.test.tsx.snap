// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField test InputField disabled state 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #F7F8FA;
  border: 0.0625rem solid #F7F8FA;
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #4475E3;
}

.c4:hover:enabled {
  border-color: #2F65E0;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: #4475E3;
}

.c4:disabled {
  color: #DBDBDB;
  border-color: rgba(219,219,219,0.2);
  background-color: rgba(219,219,219,0.2);
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid rgba(219,219,219,0.2);
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #F7F8FA !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: #4475E3;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: #2F65E0;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  gap: 0.5rem;
  height: 1.125rem;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #D14343;
  padding-left: 1rem;
  height: 1.0625rem;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
        disabled=""
      >
         
      </div>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
          disabled=""
          value="disabled"
        />
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
      >
         
      </div>
    </div>
  </div>
</body>
`;

exports[`InputField test InputField error state 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #D14343;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #FFEBEB;
  border: 0.0625rem solid #FFEBEB;
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #D14343;
}

.c4:hover:enabled {
  border-color: transparent;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: transparent;
}

.c4:disabled {
  color: #DBDBDB;
  border-color: rgba(219,219,219,0.2);
  background-color: rgba(219,219,219,0.2);
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid #FFEBEB;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #FFEBEB !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: transparent;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: transparent;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #D14343;
  gap: 0.5rem;
  height: 1.125rem;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #D14343;
  padding-left: 1rem;
  height: 1.0625rem;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
      >
         
      </div>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
        />
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
      >
         
      </div>
    </div>
  </div>
</body>
`;

exports[`InputField type and clear InputField test 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #F7F8FA;
  border: 0.0625rem solid #F7F8FA;
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #4475E3;
}

.c4:hover:enabled {
  border-color: #2F65E0;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: #4475E3;
}

.c4:disabled {
  color: #DBDBDB;
  border-color: rgba(219,219,219,0.2);
  background-color: rgba(219,219,219,0.2);
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid #4475E3;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #F7F8FA !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: #4475E3;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: #2F65E0;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  gap: 0.5rem;
  height: 1.125rem;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #D14343;
  padding-left: 1rem;
  height: 1.0625rem;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
      >
         
      </div>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
        />
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
      >
         
      </div>
    </div>
  </div>
</body>
`;
